src/plugins/ui_actions/public/actions/action.ts

```diff
@@ -17,49 +17,132 @@
  * under the License.
  */
 
-import { UiComponent } from 'src/plugins/kibana_utils/common';
-
-export interface Action<ActionContext extends {} = {}> {
-  /**
-   * Determined the order when there is more than one action matched to a trigger.
-   * Higher numbers are displayed first.
-   */
-  order?: number;
+import { Presentable } from '../util/presentable';
+import { Configurable } from '../util/configurable';
+import { ActionContract } from './action_contract';
 
+/**
+ * Legacy action interface, do not use. Use @type {ActionDefinition} and
+ * @type {ActionInternal} instead.
+ *
+ * @deprecated
+ */
+export interface Action<Context extends {} = {}> extends Partial<Presentable<Context>> {
   id: string;
 
   readonly type: string;
 
   /**
    * Optional EUI icon type that can be displayed along with the title.
    */
-  getIconType(context: ActionContext): string | undefined;
+  getIconType(context: Context): string | undefined;
 
   /**
    * Returns a title to be displayed to the user.
    * @param context
    */
-  getDisplayName(context: ActionContext): string;
-
-  /**
-   * `UiComponent` to render when displaying this action as a context menu item.
-   * If not provided, `getDisplayName` will be used instead.
-   */
-  MenuItem?: UiComponent<{ context: ActionContext }>;
+  getDisplayName(context: Context): string;
 
   /**
    * Returns a promise that resolves to true if this action is compatible given the context,
    * otherwise resolves to false.
    */
-  isCompatible(context: ActionContext): Promise<boolean>;
+  isCompatible(context: Context): Promise<boolean>;
 
   /**
    * If this returns something truthy, this is used in addition to the `execute` method when clicked.
    */
-  getHref?(context: ActionContext): string | undefined;
+  getHref?(context: Context): string | undefined;
+
+  /**
+   * Executes the action.
+   */
+  execute(context: Context): Promise<void>;
+}
+
+/**
+ * A convenience interface used to register an action.
+ */
+export interface ActionDefinition<
+  PresentationContext extends object = object,
+  ExecutionContext extends object = PresentationContext,
+  Return = Promise<void>,
+  Config extends object | undefined = undefined
+>
+  extends Partial<Presentable<PresentationContext>>,
```

Comment:
Discussed over Zoom: `Presentable` can be reused in other places, like `Trigger` and action factory. For `Configurable` it is nice to have it separate interface as that is what this PR adds - differentiates a regular action from dynamic one. But it could also be inlined in the action.